durabletask/worker.py

```diff
@@ -129,31 +132,117 @@ def add_activity(self, fn: task.Activity) -> str:
+                should invalidate the connection, while application-level errors
+                (bad requests, not found, etc.) should not.
+                """
+                error_code = rpc_error.code()  # type: ignore
+
+                # Connection-level errors that warrant invalidation
+                connection_level_errors = {
+                    grpc.StatusCode.UNAVAILABLE,        # Server down/unreachable
+                    grpc.StatusCode.DEADLINE_EXCEEDED,  # Timeout, likely network issue
+                    grpc.StatusCode.CANCELLED,          # Connection cancelled
+                    grpc.StatusCode.UNAUTHENTICATED,    # Auth failed, may need new connection
+                    grpc.StatusCode.ABORTED,            # Transaction aborted, connection may be bad
+                }
+
+                return error_code in connection_level_errors
+
             # TODO: Investigate whether asyncio could be used to enable greater concurrency for async activity
             #       functions. We'd need to know ahead of time whether a function is async or not.
-            # TODO: Max concurrency configuration settings
-            with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+            with concurrent.futures.ThreadPoolExecutor(max_workers=self._max_workers, thread_name_prefix="DurableTask") as executor:
                 while not self._shutdown.is_set():
-                    try:
-                        # send a "Hello" message to the sidecar to ensure that it's listening
-                        stub.Hello(empty_pb2.Empty())
+                    # Ensure we have a valid connection before attempting work
+                    if current_stub is None:
+                        try:
+                            create_fresh_connection()
+                        except Exception:
+                            # Connection failed, implement exponential backoff
+                            conn_retry_count += 1
+                            delay = min(conn_max_retry_delay, (2 ** min(conn_retry_count, 6)) + random.uniform(0, 1))
+                            self._logger.warning(f'Connection failed, retrying in {delay:.2f} seconds (attempt {conn_retry_count})')
+                            if self._shutdown.wait(delay):
+                                break  # Shutdown requested during wait
+                            continue
 
-                        # stream work items
+                    try:
+                        # Stream work items with the current connection
+                        # Type assertion since we know current_stub is not None at this point
+                        assert current_stub is not None, "current_stub should not be None at this point"
+                        stub = current_stub  # Local reference for type safety
                         self._response_stream = stub.GetWorkItems(pb.GetWorkItemsRequest())
```

Comment:
As noted in my other comments, we should pass the user-provided concurrency values as parameters to `GetWorkItems`. This will ensure that DTS doesn't try to send more work items than can be handled by the worker.